lib/CL/devices/pthread/pthread.c

```diff
@@ -117,6 +113,10 @@ struct data {
   mem_regions_management* mem_regions;
 #endif
 
+  /* Lock for command list related operations */
+  pthread_mutex_t cq_lock __attribute__ ((aligned (CACHELINE_SIZE)));
```

Comment:
> Volatiles should not cause accesses to main memory / cache

AFAICT, they may cause quite a lot of it. If e.g. we have a `struct { volatile int i; pthread_mutex_t lock; }` then reading `i` in a loop will cause an L3/memory fetch everytime some other thread does a pthread lock, because they share a cacheline and it's invalidated by the lock. So volatiles can in fact be very expensive.